compose/foundation/foundation/src/desktopMain/kotlin/androidx/compose/foundation/MouseDraggable.desktop.kt

```diff
@@ -0,0 +1,143 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.compose.foundation
+
+import androidx.compose.foundation.gestures.awaitPointerSlopOrCancellation
+import androidx.compose.foundation.gestures.drag
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.ExperimentalComposeUiApi
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.composed
+import androidx.compose.ui.geometry.Offset
+import androidx.compose.ui.input.pointer.PointerInputChange
+import androidx.compose.ui.input.pointer.PointerKeyboardModifiers
+import androidx.compose.ui.input.pointer.pointerInput
+import androidx.compose.ui.input.pointer.positionChange
+import androidx.compose.ui.platform.LocalKeyboardModifiers
+import kotlinx.coroutines.currentCoroutineContext
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.isActive
+
+@ExperimentalFoundationApi
+data class DragChange(
+    val offset: Offset = Offset.Zero,
+    val previousKeyboardModifiers: PointerKeyboardModifiers,
+    val currentKeyboardModifiers: PointerKeyboardModifiers
+) {
+    fun keyboardModifierChanged(): Boolean =
+        previousKeyboardModifiers != currentKeyboardModifiers
+}
+
+
+@OptIn(ExperimentalComposeUiApi::class)
+@ExperimentalFoundationApi
+/**
+ * @param pressFilter - allows to filter press events which initialise the drag. By default, it expects
+ * either [PointerButton.isPrimary] for mouse or usual tap for non-mouse pointers.
+ * @param onDrag - receives an instance of [DragChange]  for every pointer position change when dragging or
+ * when [PointerKeyboardModifiers] state changes after drag started and before it ends.
+ */
+fun Modifier.draggable(
+    enabled: Boolean = true,
+    pressFilter: ClickFilterScope.() -> Boolean = {
+        // Primary button for mouse and usual tap for non-mouse pointers
+        if (isMouse && relatedPointerButton != null) {
+            relatedPointerButton.isPrimary
+        } else {
+            allChangedToDown()
+        }
+    },
+    onDragStart: (Offset, PointerKeyboardModifiers) -> Unit = { _, _ -> },
+    onDragCancel: () -> Unit = {},
+    onDragEnd: () -> Unit = {},
+    onDrag: (DragChange) -> Unit
+): Modifier = composed {
+    val keyModifiers = LocalKeyboardModifiers.current
```

Comment:
Thanks!

>  filter uses ClickFilterScope (name might be confusing), but it seems it fits well from implementation needs perspective.

If we name it PointerFilterScope, then confusing will go away :).

Anyway, it fits well into `draggable`, I agree. `pressFilter` probably should be renamed to `filter`, because we also need to check it on the release (we can press the right button and release the left button). Also, we probably need to call `onDragEnd` if the move event no longer pass the filter, because we can stuck in the drag mode. For example, the `filter` has `keyboardModifier.isCtrl` and we don't check it on move. When we release the ctrl, `onDragEnd` won't be called, because it won't pass the filter.

> Here it uses new LocalKeyboardModifiers

Returning Flow<Value> instead of Value is unusual for CompositionLocal, and probably will be inconvenient as public API. Maybe I was wrong, and it is not a big deal, that we will see not the real keyboard modifiers at the event moment, only modifiers that were recomposed 16ms ago? Because events (Press of Ctrl) and visual representation (`Plus` icon) probably should be in sync with each other. If we consider that, we can return a value instead of Flow:
```
val LocalPointerKeyboardModifiers = staticCompositionLocalOf<PointerKeyboardModifiers> {
    noLocalProvidedFor("LocalKeyboardModifiers")
}
```

> onDrag callback will be invoked for every movement while dragging + for every LocalKeyboardModifiers state change.

If we change Flow<Value> to Value then I am a little worried that there will be two similar ways to know the current pressed modifiers:
```
val modifiers = LocalPointerKeyboardModifiers.current

Box(
    Modifier
      .draggable(
         onDragEnd = {
             if (modifiers.isCtrl) viewModel.copy() else viewModel.move()
         }
      )
      { ... }
      .pointerIcon(if (modifiers.isCtrl) PlusCursor else DefaultCursor)
)
```
```
var modifiers by remember { mutableStateOf(KeyboardModifiers()) }

Box(
    Modifier
      .draggable(
         onDragEnd = {
             if (modifiers.isCtrl) viewModel.copy() else viewModel.move()
         }
      )
      {
           modifiers = it.keyboardModifiers
      }
      .pointerIcon(if (modifiers.isCtrl) PlusCursor else DefaultCursor)
)
```

Maybe if we get rid of `Flow`, we can keep only `LocalPointerKeyboardModifiers`? Also, the `filter` parameter probably should stay, because we can use a different logic for starting/finish a drag: drag starts with pressed Shift, and Ctrl we use only for additional logic during the drag.

Reply:
```kotlin
val modifiers = LocalPointerKeyboardModifiers.current

Box(
    Modifier
      .draggable(
         onDragEnd = {
             if (modifiers.isCtrl) viewModel.copy() else viewModel.move()
         }
      )
      { ... }
      .pointerIcon(if (modifiers.isCtrl) PlusCursor else DefaultCursor)
)
```

It seems that LocalPointerKeyboardModifiers (with value) can be changed quite frequently  (comparing with other Composition Locals). And it can be read in multiple places (example: many draggable items). 
So every time it changes, every scope reading its value will be recomposed. Even when there is no drag in progress. In our use case (for pointerIcon), such a recomposition will be rather useless when there is no drag in progress (or when item is not hovered at least). 
I think we would like to avoid unnecessary recompositions. And it seems reading such a volatile state as LocalPointerKeyboardModifiers might be not good. Do you also see such a risk? Or do I exaggerate the potential issue?

Reply:
I haven't thought about it 🤔. It is a valid concern.

But it seems the difference between composition local and the other ways to listen only in the ability to check condition earlier, before the recomposition:

not
```
var isDragging by ...
var isCtrl by ...

LaunchedEffect {
   keyModifiers.collect {
     isCtrl = it.isCtrl
  }
}

// or
Box(Modifier.onKeyModifierChanged { 
  isCtrl = it.isCtrl
})

if (isDragging && isCtrl) { ... }
```
but
```
var isDragging by ...
var isCopy by ...

LaunchedEffect {
   keyModifiers.collect {
     isCopy = isDragging && it.isCtrl
  }
}

// or
Box(Modifier.onKeyModifierChanged { 
  isCopy = isDragging && it.isCtrl
})

if (isDragging && isCopy) { ... }
```

After I wrote this code, I realised that this won't work if we press Ctrl before the drag (listener won't be triggered).

I think, to simplify the code and fix the issue, we have to set the variable unconditionally, and check the condition on the level of composition. If we do that, it will probably be the same as we read CompositionLocal. And if we want to optimize the recomposition, we can just move code into a separate function. But if CompositionLocal behaves differently from mutableStateOf, and we recompose the entire tree (not only functions where we read the CompositionLocal) then CompositionLocal is definitely not an option.

Reply:
We decided to try adding `keyboardModifiers` into `WindowInfo` (it can be accessed via `LocalWindowInfo`).
 `keyboardModifiers` is implemented as MutableState.